### ISSUE TITLE
Issue regarding rlhf implementation

### ISSUE THREAD

**Issue by Pooja-1410** [ashishjamarkattel/reinforment-learning-with-human-feedback#2]
from random import choices
from tqdm import tqdm
import time
import numpy as np
import ast

for epoch in range(1):
    for batch in tqdm(ppo_trainer.dataloader):
        (logs, game_data,) = (
            dict(),
            dict(),
        )

        print(ctrl_str)
        #### prepend a random control token
        task_list = choices(ctrl_str, k=config.batch_size)
    # k
        # task_list_tensors = [torch.tensor(eval(item)) if isinstance(item, str) else torch.tensor(item) for item in task_list]

# Move each tensor to CUDA if a GPU is available
        # task_list_tensors_cuda = [tensor.cuda() if torch.cuda.is_available() else tensor for tensor in task_list_tensors]
        game_data["query"] = [t + q for t, q in zip(task_list, batch["query"])]
        cuda = torch.device("cuda")
        query_tensors = [torch.cat((ctrl_tokens[t].to(cuda), input_ids.to(cuda))) for t, input_ids in zip(task_list, batch["input_ids"])]

        #### get response from gpt2
        response_tensors = []
        for query in query_tensors:
            response = ppo_trainer.generate(query.to(cuda), **generation_kwargs)
            response_on_device = response.to(cuda)
            response_tensors.append(response_on_device.squeeze()[-txt_out_len:])
#         print(response_tensors)
        game_data["response"] = [starcoder_tokenizer.decode(r.squeeze()) for r in response_tensors]

        #### sentiment analysis
        texts = [q + r for q, r in zip(batch["query"], game_data["response"])]
        logits = get_score(starcoder_model, starcoder_tokenizer, texts)
        logits = [logit.cuda() if torch.cuda.is_available() else logit for logit in logits]
        rewards = pos_logit_to_reward(logits, task_list)
        print(rewards)
        #logits = get_score(starcoder_model,starcoder_tokenizer, texts)
        #logits = logits.cuda() if torch.cuda.is_available() else logits
        #rewards = pos_logit_to_reward(logits, task_list)

        #### Run PPO training
        t = time.time()
        
        stats = ppo_trainer.step(query_tensors, response_tensors, rewards)

        for cs in ctrl_str:
            key = "env/reward_" + cs.strip("[]")
            stats[key] = np.mean([r.cpu().numpy() for r, t in zip(rewards, task_list) if t == cs])
        ppo_trainer.log_stats(stats, game_data, rewards)

the above is the code snippet and we have attached the error image 
![Screenshot (113)](https://github.com/ashishjamarkattel/reinforment-learning-with-human-feedback/assets/91204961/a69e4ad4-7050-4661-b8e2-c65cf6c3a326)

**Comment by ashishjamarkattel** [comment]
Seems you are having hard time keeping everything in cuda, 
for the step try consider keeping query_tensor, response_tensor, and reward all in cuda
before the ppo_trainer.step
add query_tensors = query_tensor.to(cuda)
response_tensor = response_tensor.to(cuda)
reward = response_tensor.to(cuda)

` it should solve the issue. Let me know if problem exit. p.s i tried running same code, didnot face this issue.`

**Comment by Pooja-1410** [comment]
We are running the code on T4 GPU on which one you are running ?

**Comment by ashishjamarkattel** [comment]
I used the free version which should be t4. Is your problem still not solved?

**Comment by Pooja-1410** [comment]
Yeah it's still not working. 🙃

**Comment by Pooja-1410** [comment]
Are there steps that we are messing up. Is there any alternative solution

**Comment by ashishjamarkattel** [comment]
I hope not,
Can you share me the dataset and reward model so i can try to replicate it

**Comment by Pooja-1410** [comment]
Yeah, we have used same dataset that you did
[dataset.zip](https://github.com/ashishjamarkattel/reinforment-learning-with-human-feedback/files/13614971/dataset.zip)
[reward dataset.zip](https://github.com/ashishjamarkattel/reinforment-learning-with-human-feedback/files/13614974/reward.dataset.zip)

**Comment by ashishjamarkattel** [comment]
Is your problem solved?
Please reinitialize the gpu after every step it would solve your issue?

**Comment by Pooja-1410** [comment]
No that doesn't solve problem reinitializing everytime tends to lose the folders that we create at step 1 .So, it doesn't proceed
.Is it possible for you to have a google meet with our team so that we can show you the problem.

**Comment by ashishjamarkattel** [comment]
I understand the problem and sorry could not make it ,Currently occupied with multiple task. I think you need to save the model to drive after training each step. and take the model each time. without problem of you training again and again that way it would be much faster. 

or make the individual .ipynb file for each step.

The problem exist because of some cached variable that exists which is being referenced from step 1 or 2 to step 3.

someone solved the issue same as yours
![image](https://github.com/ashishjamarkattel/reinforment-learning-with-human-feedback/assets/63300562/050b105a-8ab7-40d9-86b9-077baab6e768)

![image](https://github.com/ashishjamarkattel/reinforment-learning-with-human-feedback/assets/63300562/bad87534-d8f7-4a2e-9d9f-25c804e5605d)

just running individual step.


let me know if the problem still exist, i will try to

**Comment by Pooja-1410** [comment]
Thank you so much for your guidance , the above image has helped a lot .
There are some things we are doubtful about like
1) Code is working successfully but how we are showing the input ouputs .
2)How exactly the reward modeling is working with text .
Can you help us understand that?

**Comment by ashishjamarkattel** [comment]
Think reward model as the penality function. It calculate the similarity score between human written text and actual text.
so in the training of step 3, we tends to bring the generated text much similar to human written text, by introducing the penality.

**Comment by Pooja-1410** [comment]
In code where is the input and output part like summarized text as in our case
